types/lib/mapper.go

```diff
@@ -0,0 +1,276 @@
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.ElemKey(index))
+	return lm.cdc.UnmarshalBinary(bz, ptr)
+}
+
+func (lm GenericMapper) Set(ctx sdk.Context, index uint64, value interface{}) {
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.ElemKey(index), bz)
+}
+
+func (lm GenericMapper) Delete(ctx sdk.Context, index uint64) {
+	store := ctx.KVStore(lm.key)
+	store.Delete(lm.ElemKey(index))
+}
+
+func (lm GenericMapper) Push(ctx sdk.Context, value interface{}) {
+	length := lm.Len(ctx)
+	lm.Set(ctx, length, value)
+
+	store := ctx.KVStore(lm.key)
+	store.Set(lm.LengthKey(), marshalUint64(lm.cdc, length+1))
+}
+
+func (lm GenericMapper) GetMeta(ctx sdk.Context, ptr interface{}) error {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.MetaKey())
+	return lm.cdc.UnmarshalBinary(bz, ptr)
+}
+
+func (lm GenericMapper) SetMeta(ctx sdk.Context, value interface{}) {
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.MetaKey(), bz)
+}
+
+func (lm GenericMapper) IterateRead(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, uint64) bool) {
+	store := ctx.KVStore(lm.key)
+	start, end := subspace([]byte(fmt.Sprintf("%s/elem/", lm.prefix)))
+	iter := store.Iterator(start, end)
+	for ; iter.Valid(); iter.Next() {
+		v := iter.Value()
+		if err := lm.cdc.UnmarshalBinary(v, ptr); err != nil {
+			panic(err)
+		}
+		s := strings.Split(string(iter.Key()), "/")
+		index, err := strconv.ParseUint(s[len(s)-1], 10, 64)
+		if err != nil {
+			panic(err)
+		}
+		if fn(ctx, index) {
+			break
+		}
+	}
+
+	iter.Close()
+}
+
+func (lm GenericMapper) IterateWrite(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, uint64) bool) {
+	length := lm.Len(ctx)
+
+	for i := uint64(0); i < length; i++ {
+		if err := lm.Get(ctx, i, ptr); err != nil {
+			continue
+		}
+		if fn(ctx, i) {
+			break
+		}
+	}
+}
+
+func (lm GenericMapper) LengthKey() []byte {
+	return []byte(fmt.Sprintf("%s/length", lm.prefix))
+}
+
+func (lm GenericMapper) ElemKey(i uint64) []byte {
+	return []byte(fmt.Sprintf("%s/elem/%020d", lm.prefix, i))
+}
+
+func (lm GenericMapper) MetaKey() []byte {
+	return []byte(fmt.Sprintf("%s/meta", lm.prefix))
+}
+
+// QueueMapper is a Mapper interface that provides queue-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+
+type QueueMapper interface {
+	Push(sdk.Context, interface{})
+	// Popping/Peeking on an empty queue will cause panic
+	// The user should check IsEmpty() before doing any actions
+	Peek(sdk.Context, interface{}) error
+	Pop(sdk.Context)
+	IsEmpty(sdk.Context) bool
+	// Iterate() removes elements it processed
+	// Return true in the continuation to break
+	// The interface{} is unmarshalled before the continuation is called
+	// Starts from the top(head) of the queue
+	Flush(sdk.Context, interface{}, func(sdk.Context) bool)
```

Comment:
Unsafe to perform `Pop()` or `Push()` in the continuation presumably?

Needs a `Contract: ...` line